sentinel-core/log/metric_writer.cc

```diff
@@ -0,0 +1,280 @@
+#include "sentinel-core/log/metric_writer.h"
+
+#include <dirent.h>
+#include <unistd.h>
+
+#include <ctime>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <locale>
+#include <regex>
+#include <sstream>
+
+#include "absl/strings/match.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+
+#include "sentinel-core/log/log_base.h"
+#include "sentinel-core/log/record_log.h"
+
+#include "sentinel-core/utils/file_utils.h"
+#include "sentinel-core/utils/time_utils.h"
+
+using namespace Sentinel::Utils;
+
+namespace Sentinel {
+namespace Log {
+
+const std::string MetricWriter::kMetricFile = "metrics.log";
+const std::string MetricWriter::kMetricIndexFileSuffix = ".idx";
+
+MetricWriter::MetricWriter(int64_t single_file_size, int32_t total_file_count)
+    : single_file_size_(single_file_size), total_file_count_(total_file_count) {
+  RecordLog::Info("[MetricWriter] Creating new MetricWriter, singleFileSize=" +
+                  std::to_string(single_file_size_) +
+                  ", totalFileCount=" + std::to_string(total_file_count_));
+
+  base_dir_ = LogBase::GetLogBaseDir();
+
+  auto time = TimeUtils::CurrentTimeMillis();
+  last_second_ = time.count() / 1000;
+
+  single_file_size_ = single_file_size;
+  total_file_count_ = total_file_count;
+
+  std::tm tm = {};
+  std::istringstream ss("1970-01-01 00:00:00");
+  ss >> std::get_time(&tm, "%Y-%m-%d T %H:%M:%S");
+  auto tp = std::chrono::system_clock::from_time_t(std::mktime(&tm));
+  time_second_base_ =
+      std::chrono::duration_cast<std::chrono::seconds>(tp.time_since_epoch())
+          .count();
+
+  pid_ = ::getpid();
+}
+
+/**
+ * 如果传入了time，就认为nodes中所有的时间时间戳都是time.
+ *
+ * @param time
+ * @param nodes
+ */
+void MetricWriter::Write(int64_t time, std::vector<Metric::MetricNode> &nodes) {
+  std::lock_guard<std::mutex> lk(lock_);
+
+  if (time != -1) {
+    for (auto &node : nodes) {
+      node.SetTimestamp(time);
+    }
+  }
+
+  auto app_name = Config::LocalConfigSingleton().get().app_name();
+
+  // first write, should create file
+  if (!metric_out_.is_open()) {
+    base_file_name_ = FormMetricFileName(app_name, pid_);
+    CloseAndNewFile(NextFileNameOfDay(time));
+  }
+
+  auto second = time / 1000;
+
+  if (second < last_second_) {
+    // 时间靠前的直接忽略，不应该发生。
+  } else if (second == last_second_) {
+    DoWrite(time, nodes);
+  } else {
+    WriteIndex(second, metric_out_.tellp());
+    if (IsNewDay(last_second_, second)) {
+      CloseAndNewFile(NextFileNameOfDay(time));
+      DoWrite(time, nodes);
+    } else {
+      DoWrite(time, nodes);
+    }
+    last_second_ = second;
+  }
+}
+
+void MetricWriter::DoWrite(int64_t time,
+                           std::vector<Metric::MetricNode> &nodes) {
+  for (auto &node : nodes) {
+    metric_out_ << node.ToFatString();
```

Comment:
We may also need to write `\n` per line.

```suggestion
    metric_out_ << node->ToFatString() << "\n";
```